rust-utils/src/conductor_config.rs

```diff
@@ -8,21 +8,117 @@ use holochain_p2p::kitsune_p2p::{
   dependencies::kitsune_p2p_types::config::tuning_params_struct::KitsuneP2pTuningParams,
   KitsuneP2pConfig, TransportConfig,
 };
-use std::path::Path;
+use napi::{Error, Result, Status};
+use napi_derive::napi;
+use serde_yaml::{Mapping, Value};
+use std::path::{Path, PathBuf};
+
+fn create_error(msg: &str) -> Error {
+  Error::new(Status::GenericFailure, String::from(msg))
+}
+
+fn insert_mapping(mapping: &mut Mapping, key: &str, value: Value) {
+  mapping.insert(Value::String(String::from(key)), value);
+}
+
+fn create_mapping_with_entries(entries: Vec<(&str, Value)>) -> Mapping {
+  let mut mapping = Mapping::new();
+  for (key, value) in entries {
+    insert_mapping(&mut mapping, key, value);
+  }
+  mapping
+}
+
+#[napi]
+pub fn overwrite_config(
+  admin_port: u16,
+  keystore_connection_url: String,
+  bootstrap_server_url: Option<String>,
+  signaling_server_url: Option<String>,
+  config_path: String,
+) -> Result<String> {
+  let mut config = std::fs::read_to_string(&PathBuf::from(config_path))
+    .map_err(|_| create_error("Failed to read file"))
+    .and_then(|contents| {
+      serde_yaml::from_str::<Value>(&contents).map_err(|_| create_error("Failed to parse YAML"))
+    })
+    .and_then(|yaml| {
+      yaml
+        .as_mapping()
+        .cloned()
+        .ok_or_else(|| create_error("Expected YAML content to be a mapping"))
+    })?;
+
+  let websocket_interface = create_mapping_with_entries(vec![
+    ("type", Value::String(String::from("websocket"))),
+    ("port", Value::Number(admin_port.into())),
+  ]);
+
+  let admin_interface =
+    create_mapping_with_entries(vec![("driver", Value::Mapping(websocket_interface))]);
+
+  insert_mapping(
+    &mut config,
+    "admin_interfaces",
+    Value::Sequence(vec![Value::Mapping(admin_interface)]),
+  );
+
+  insert_mapping(
+    &mut config,
+    "keystore",
+    Value::Mapping(create_mapping_with_entries(vec![
+      ("type", Value::String(String::from("lair_server"))),
+      ("connection_url", Value::String(keystore_connection_url)),
+    ])),
+  );
+
+  let network = config
+    .get_mut(&Value::String(String::from("network")))
+    .and_then(|v| v.as_mapping_mut())
+    .ok_or_else(|| create_error("Expected 'network' entry in the config"))?;
+
+  bootstrap_server_url.map(|url| {
```

Comment:
The bootstrap and signaling URLs may also be None and in this case nothing should be overwritten. Looking at the code here (without having it tested) it seems that even if it is None, the value in the existing config will be overwritten which should not be the case.

Reply:
from what I've read, map will execute only if value is not None, so it will remain unchanged

Reply:
Oh right, of course.